File: html/old/R2/SACO R2 2010 (Solutions)/Mansion.java
import java.util.*;

public class Mansion {

    public static void main (String[] args) {
        Scanner sc = new Scanner(System.in);
        System.out.print("How many rows and columns are there? ");
        int rows = sc.nextInt();
        int columns = sc.nextInt();

        String[] map = new String[rows];
        Set<Coord> lightsFound = new TreeSet<Coord>();
        Queue<Coord> newLights = new LinkedList<Coord>();
        System.out.println("Enter the mansion one row per line:")
        for (int r = 0; r < rows; ++r) {
            map[r] = sc.next();
            for (int c = 0; c < columns; ++c) {
                if (map[r].charAt(c) == 'S') {
                    Coord start = new Coord(r,c);
                    lightsFound.add(start);
                    newLights.add(start);
                }
            }
        }

        int[] stepx = {-1, 0, 1, 0, 0};
        int[] stepy = {0, -1, 0, 1, 0};
        int[] lightx = {-1, -1, 0, 1, 1, 1, 0, -1, 0};
        int[] lighty = {0, -1, -1, -1, 0, 1, 1, 1, 0};

        while (true) {
            Coord light = newLights.poll();
            if (light == null)
                break;

            int r = light.row;
            int c = light.column;

            for (int i = 0; i < lightx.length; ++i) {
                for (int j = 0; j < stepx.length; ++j) {
                    for (int k = 0; k < stepx.length; ++k) {
                        int r1 = r + lighty[i] + stepy[j] + stepy[k];
                        int c1 = c + lightx[i] + stepx[j] + stepx[k];
                        Coord coord = new Coord(r1, c1);

                        if (0 <= r1 && r1 < rows && 0 <= c1 && c1 < columns
                            && map[r1].charAt(c1) == '*' &&
                            !lightsFound.contains(coord)) {

                            lightsFound.add(coord);
                            newLights.add(coord);
                        }
                    }
                }
            }
        }

        System.out.println("I can switch on " + (lightsFound.size()-1)
                            + " lights.");
    }
}

class Coord implements Comparable<Coord> {
    final int row, column;
    public Coord (int r, int c) {
        row = r;
        column = c;
    }

    public int compareTo (Coord other) {
        if (row != other.row)
            return row - other.row;
        else
            return column - other.column;
    }
}

File: html/old/R2/SACO R2 2010 (Solutions)/mansion.py
#!/usr/bin/env python
# Keegan Carruthers-Smith - Mansion - SACO R2 2010

line = raw_input('How many rows and columns are there? ')
rows, cols = map(int, line.split())

print 'Enter the mansion one row per line:'
mansion = []
for r in range(rows):
    mansion.append(list(raw_input()))
    if 'S' in mansion[-1]:
        row, col = r, mansion[-1].index('S')

delta_walk = [(r, c) for r in range(-2, 3) for c in range(-2, 3)
              if 0 < abs(r) + abs(c) <= 2]
delta_light = [(r, c) for r in range(-1, 2) for c in range(-1, 2)]
valid_coord = lambda r, c : 0 <= r < rows and 0 <= c < cols

lights_switched = 0
seen = set([(row, col)])
queue = [(row, col)]
while queue:
    row, col = queue.pop()

    if mansion[row][col] in 'S*':
        for r, c in delta_light:
            r, c = row + r, col + c
            if valid_coord(r, c) and mansion[r][c] == '.':
                mansion[r][c] = '$'
        if mansion[row][col] == '*':
            lights_switched += 1
        mansion[row][col] = '$'

    for r, c in delta_walk:
        r, c = row + r, col + c
        if valid_coord(r, c) and (r,c ) not in seen and mansion[r][c] != '.':
            seen.add((r, c))
            queue.append((r, c))

print 'I can switch on %d lights.' % lights_switched
